src/commands/math/commands.js

```diff
@@ -1007,6 +1007,58 @@ Environments.matrix = P(Environment, function(_, super_) {
     table.toggleClass('mq-rows-1', table.find('tr').length === 1);
     this.relink();
   };
+  _.upInto = function(dir, cursor) {
+    var rowSize = this.getRowSize();
```

Comment:
It could be more performant to calculate and cache the rowSize when the matrix is created or edited, as it otherwise won't change.

Maybe during `this.finalizeTree()`.